### PR TITLE
APPS-1104 Feature: Bugsnag implementation to track bugs

### PR DESCRIPTION
**Pull request by multinauta**
**Refers:** 
- https://shipitchile.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=APPS&modal=detail&selectedIssue=APPS-1104
- https://shipitchile.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=APPS&modal=detail&selectedIssue=APPS-1209
- https://shipitchile.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=APPS&modal=detail&selectedIssue=APPS-1111

**Why I'm adding these lines of code?**

- I'm implementing bugsnag so that I can catch the exceptions of the Web Services from the store side
- I'm fixing the way to get the measurements sent
- I'm improvement the way to catch the exceptions
- I added the option to be able to select all the communes in the shipping configurations
- Add a parameterizable cron from the shipping configuration, which, according to the selected frequency, forwards the orders that could not be sent on the first attempt
- I created 2 states 'Cancelado Shipit' and 'Otro Shpit'', which validate the states 'Cancelado'  and 'Otros', and pending is validated as 'El envío no fue retirado', so as not to use states that reverse inventory
- fix logo on shipit shipping method

**Tests**

**Shipping test:** test of sending a sale.
sent_test

----------------------------------------------------------------------------------
**Bugsnag Test:** bugsnag catching the exception.

bugsnag_test

----------------------------------------------------------------------------------
**Select all communes test:** through jquery all communes are selected.
test_communes_1
test_communes_2

----------------------------------------------------------------------------------
**complementary shipping:** configuration of the forwarding frequency of records that could not be sent in the first instance.

cron_config

----------------------------------------------------------------------------------
**validation of states that do not reverse inventory: ** test in which each of the three states that reverted the inventory is updated and that are now validated.
status

**Deployment**

- Merge and create a release
- publicate publish on the wordpress marketplace
- approve and send to production this pr https://github.com/shipit-team/orders/pull/199

### CODE REVIEW

**Comment by Zeaneth** [comment]
Maybe add a Tooltip to the "Todas las comunas?" and "Frecuencia" options? I think they're confusing without an additional explanation.

File: shipit/trunk/src/webhook.php [hunk 1]
```diff
@@ -184,6 +187,33 @@ function shipit_register_my_new_order_statuses() {
       'show_in_admin_status_list' => true,
       'label_count' => _n_noop('Devuelto <span class="count">(%s)</span>', 'Devuelto<span class="count">(%s)</span>', 'woocommerce')
     ));
+    register_post_status('wc-withdrawn', array(
+      'label' => _x('withdrawn', 'Order status', 'woocommerce'),
+      'public' => true,
+      'exclude_from_search' => false,
+      'show_in_admin_all_list' => true,
+      'show_in_admin_status_list' => true,
+      'label_count' => _n_noop('El envio no fue retirado <span class="count">(%s)</span>', 'El envio no fue retirado<span class="count">(%s)</span>', 'woocommerce')
```
Comment (tostitos-dev) [inline shipit/trunk/src/webhook.php:196]:
```suggestion
      'label_count' => _n_noop('El envío no fue retirado <span class="count">(%s)</span>', 'El envío no fue retirado<span class="count">(%s)</span>', 'woocommerce')
```

File: shipit/tags/3.1.0/shipit.php [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+              'css' => 'max-width:200px;',
+            ),
+            'all_free_communes' => array(
+              'title' => __('Todas las comunas?', 'dc_raq'),
+              'type' => 'checkbox',
+              'description' => __('Todas las comunas?', 'dc_raq'),
+              'default' => 'no'
+            ),
+            'free_communes' => array(
+              'title' => __('Comunas especificas', 'woocommerce'),
```
Comment (tostitos-dev) [inline shipit/tags/3.1.0/shipit.php:319]:
```suggestion
              'title' => __('Comunas específicas', 'woocommerce'),
```

File: shipit/tags/3.1.0/shipit.php [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+              'class' => 'wc-enhanced-select',
+              'options' => WC()->countries->get_states('CL'),
+              'custom_attributes' => array(
+                'data-placeholder' => __('Seleccione comunas', 'woocommerce'),
+              ),
+            ),
+            'price' => array(
+              'title' => __('Envios gratis a partir:', 'woocommerce'),
+              'type' => 'number',
+              'description' => __('Configure el valor de minimo de orden para despachos.', 'woocommerce'),
```
Comment (tostitos-dev) [inline shipit/tags/3.1.0/shipit.php:331]:
```suggestion
              'description' => __('Configure el valor de mínimo de orden para despachos.', 'woocommerce'),
```

File: shipit/tags/3.1.0/src/shipit-settings.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+              'default' => get_option('shipit_token')
+            )
+          )
+        );
+        return $settings_fields;
+      }
+
+      function plugin_page() {
+        echo '<div class="wrap">';
+        echo'<h2>Inicio de sesion de usuario</h2> Bienvenido a la configuración de plugin Shipit para Woocommerce </div>';
```
Comment (tostitos-dev) [inline shipit/tags/3.1.0/src/shipit-settings.php:59]:
```suggestion
        echo'<h2>Inicio de sesión de usuario</h2> Bienvenido a la configuración de plugin Shipit para Woocommerce </div>';
```

File: shipit/tags/3.1.0/src/shipit_service/address.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    function getLatitude() {
+      return $this->latitude;
+    }
+
+    function getLongitude() {
+      return $this->longitude;
+    }
+  }
+?>
```
Comment (tostitos-dev) [inline shipit/tags/3.1.0/src/shipit_service/address.php:48]: EOF

File: shipit/tags/3.1.0/src/auther.php [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+    global $wp_json_basic_auth_error;
+    return $wp_json_basic_auth_error;
+  }
+
+  add_filter('rest_authentication_errors', 'shipit_json_basic_auth_error');
+  add_filter('woocommerce_states', 'chileanDestinies');
+
+  function chileanDestinies($states) {
+    $states['CL'] = array(
+      'CL410' => __("ACHAO", 'woocommerce'),
```
Comment (tostitos-dev) [inline shipit/tags/3.1.0/src/auther.php:36]: Are these communes different from those obtained by the API?

Comment (Zeaneth) [reply shipit/tags/3.1.0/src/auther.php:36]: They're the same, but if we get them via API each time a customer requires a shipping rate, our response time is really low!

File: shipit/trunk/src/shipit_service/bugsnag.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    public function bugsnagLog($response, $from, $reference) {
+      if ((isset($response->status)) && ($response->status != 200)) {
+        $r = $this->sendNotification("code error ".$response->status.", name error ".$response->error.", from ".$from.": references : ".$reference, $response->exception);
+        ShipitDebug::debug("code error ".$response->status.", name error ".$response->error.", from ".$from.": ".$response->exception."la referencia ".$reference);
+      }    
+    }
+
+    public function bugsnagWpLog($response, $from, $reference) {
+      $r = $this->sendNotification("error ", $response->errors['http_request_failed'][0].", from ".$from." ,references:".$reference );
```
[the remark this answers is not in the document]
Comment (multinauta) [reply shipit/trunk/src/shipit_service/bugsnag.php]: i stay from when I was routing, I'll fix it

**Review by Zeaneth — APPROVED** [review]

**Review by FelipeChibey — APPROVED** [review]

**Review by luispascuali — APPROVED** [review]

**Pull request merged**